internal/providers/docker/components/container/build.go

```diff
@@ -91,13 +94,39 @@ func (c *Container) buildImage(ctx context.Context) error {
 		resp, err := c.Docker.ImageBuild(ctx, buildContext, opts)
 		fmt.Println("after build")
 		if resp.Body != nil {
+			defer resp.Body.Close()
 			// TODO [DOCKER_PROGRESS]: Capture progress.
-			_, _ = io.Copy(os.Stdout, resp.Body)
-			_ = resp.Body.Close()
+			//_, _ = io.Copy(os.Stdout, resp.Body)
+			scanner := bufio.NewScanner(resp.Body)
+			for scanner.Scan() {
+				var d struct {
+					ErrorDetail struct {
+						Code    int    `json:"code"`
+						Message string `json:"message"`
+					} `json:"errorDetail"`
+					Aux struct {
+						ID string `json:"ID"`
+					} `json:"aux"`
+				}
+				err := json.Unmarshal(scanner.Bytes(), &d)
+				if err != nil {
+					return err
+				}
+				if d.ErrorDetail.Message != "" {
+					return fmt.Errorf("docker build error: " + d.ErrorDetail.Message)
+				}
+				if strings.HasPrefix(d.Aux.ID, "sha256:") {
+					c.ImageID = d.Aux.ID
+					c.Image = d.Aux.ID
```

Comment:
This is `c.Spec.Image`. The spec should not be mutated as part of the controller's behavior, as those changes are not saved anywhere & are expected to be managed by the user's manifest. Maybe we should start being explicit about `c.State.Whatever` to clarify.

Reply:
Ah, that makes sense. I think I agree that giving those fields explicit names would be better.